pkg/controller/tas/topology_controller.go

```diff
@@ -0,0 +1,167 @@
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	configapi "sigs.k8s.io/kueue/apis/config/v1beta1"
+	kueuealpha "sigs.k8s.io/kueue/apis/kueue/v1alpha1"
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/cache"
+	"sigs.k8s.io/kueue/pkg/controller/core"
+	"sigs.k8s.io/kueue/pkg/queue"
+)
+
+const (
+	updateChBuffer = 10
+)
+
+type topologyReconciler struct {
+	client                client.Client
+	queues                *queue.Manager
+	cache                 *cache.Cache
+	tasCache              *cache.TASCache
+	topologyUpdateCh      chan event.GenericEvent
+	resourceFlavorHandler *resourceFlavorHandler
+}
+
+var _ reconcile.Reconciler = (*topologyReconciler)(nil)
+var _ predicate.Predicate = (*topologyReconciler)(nil)
+
+func newTopologyReconciler(c client.Client, queues *queue.Manager, cache *cache.Cache) *topologyReconciler {
+	return &topologyReconciler{
+		client:                c,
+		queues:                queues,
+		cache:                 cache,
+		tasCache:              cache.TASCache(),
+		topologyUpdateCh:      make(chan event.GenericEvent, updateChBuffer),
+		resourceFlavorHandler: &resourceFlavorHandler{},
+	}
+}
+
+func (r *topologyReconciler) setupWithManager(mgr ctrl.Manager, cfg *configapi.Configuration) (string, error) {
+	return TASTopologyController, ctrl.NewControllerManagedBy(mgr).
+		Named(TASTopologyController).
+		For(&kueuealpha.Topology{}).
+		WithOptions(controller.Options{NeedLeaderElection: ptr.To(false)}).
+		Watches(&kueue.ResourceFlavor{}, r.resourceFlavorHandler).
+		WithEventFilter(r).
+		Complete(core.WithLeadingManager(mgr, r, &kueuealpha.Topology{}, cfg))
+}
+
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=topology,verbs=get;list;watch;update
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=topology/finalizers,verbs=update
+
+func (r topologyReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	topology := &kueuealpha.Topology{}
+	if err := r.client.Get(ctx, req.NamespacedName, topology); err != nil {
+		// we'll ignore not-found errors, since there is nothing to do.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	log := ctrl.LoggerFrom(ctx).WithValues("name", req.NamespacedName.Name)
+	log.V(2).Info("Reconcile Topology")
+
+	if !topology.DeletionTimestamp.IsZero() {
+		if controllerutil.ContainsFinalizer(topology, kueue.ResourceInUseFinalizerName) {
+			if flRefs := r.tasCache.ResourceFlavorsUsingTopology(kueue.TopologyReference(topology.Name)); len(flRefs) != 0 {
```

Comment:
Can we just use `for flName, flCache := range r.tasCache.Clone() {` instead of the new function?

Reply:
Done